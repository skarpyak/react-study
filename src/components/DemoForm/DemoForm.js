import React, { Component } from 'react';
import './DemoForm.css';

class DemoForm extends Component {
    constructor (props) {
        super(props);
        this.state = {
            email: '',
            password: '',
            formErrors: {email: '', password: ''},
            emailValid: false,
            passwordValid: false,
            formValid: false,
            touched: { email: false, password: false }
        }

        this.onSignUp = this.onSignUp.bind(this);
    }

    handleUserInput (e) {
        const name = e.target.name;
        const value = e.target.value;
        this.setState({[name]: value},
            () => { this.validate() }
        );
    }

    validate() {
        const { formErrors } = this.state;

        let fieldValidationErrors = formErrors;
        let emailValid = this.state.emailValid;
        let passwordValid = this.state.passwordValid;

        emailValid = this.state.email.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
        fieldValidationErrors.email = emailValid ? '' : ' is invalid';

        passwordValid = this.state.password.length >= 6;
        fieldValidationErrors.password = passwordValid ? '': ' is too short';

        this.setState({formErrors: fieldValidationErrors,
            emailValid: emailValid,
            passwordValid: passwordValid
        }, this.validateForm);
    }

    validateForm() {
        this.setState({formValid: this.state.emailValid && this.state.passwordValid});
    }

    errorClass(error) {
        return(error.length === 0 ? '' : 'has-error');
    }

    handleInputOnBlur(e) {
        const { touched } = this.state;
        const name = e.target.name;
        this.setState({touched: { ...touched, [name]: true }},
            () => { this.validate() }
        );
    }

    onSignUp() {
        const { touched } = this.state;
        Object.keys(touched).forEach( key => touched[key] = true);
        this.setState({ touched: { ...touched }},
            this.validate()
        );

        if(this.state.formValid) {
            // sign up code here
        }

    }

    render () {
        return (
            <div className="demoForm">
                <h2>Sign up</h2>
                <div className={`form-group ${this.errorClass(this.state.formErrors.email)}`}>
                    <label htmlFor="email">Email address</label>
                    <input type="email" className="form-control"
                           name="email" value={this.state.email}
                           onChange={(event) => this.handleUserInput(event)}
                           onBlur={(event) => this.handleInputOnBlur(event) }
                    />
                    <p className={ this.state.touched.email ? '' : 'hidden' } >{this.state.formErrors.email}</p>
                </div>
                <div className={`form-group ${this.errorClass(this.state.formErrors.password)}`}>
                    <label htmlFor="password">Password</label>
                    <input type="password" className="form-control"
                           name="password"  value={this.state.password}
                           onChange={(event) => this.handleUserInput(event)}
                           onBlur={(event) => this.handleInputOnBlur(event) }
                    />
                    <p className={ this.state.touched.password ? '' : 'hidden' } >{this.state.formErrors.password}</p>
                </div>
                <button className="btn btn-primary" onClick={this.onSignUp} >
                    Sign up
                </button>
            </div>
        )
    }
}
export default DemoForm;