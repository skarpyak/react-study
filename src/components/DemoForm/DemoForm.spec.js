import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import DemoForm from './DemoForm';

let props;
const fieldChange = jest.fn();
const formChange = jest.fn();

describe('DemoForm', () => {
  beforeEach(() => {
    props = {
      state: {
        email: '',
        password: '',
        formErrors: {email: '', password: ''},
        emailValid: false,
        passwordValid: false,
        formValid: false,
        touched: {email: false, password: false}
      },
      fieldChange: fieldChange,
      formChange: formChange
    }
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<DemoForm { ...props } />, div);
  });

  it('should call fieldChange when input value changed', () => {
    const wrapper = shallow(
      <DemoForm { ...props } />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.find('[name="password"]').simulate('change', {target: {value: 'Hello', name: 'password'}});
    expect(fieldChange).toBeCalledWith({name: "password", value: "Hello"});
  });

  it('should call formChange when entering input', () => {
    const wrapper = shallow(
      <DemoForm { ...props } />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.find('[name="password"]').simulate('blur', {target: {name: 'password'}});
    expect(formChange).toBeCalledWith({"touched": {"email": false, "password": true}});
  });

  it('should call formChange when buuton clicked', () => {
    const wrapper = shallow(
      <DemoForm { ...props } />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.find('button').simulate('click', {target: {name: 'password'}});
    expect(formChange).toBeCalledWith({"touched": {"email": true, "password": true}});
  });
});

