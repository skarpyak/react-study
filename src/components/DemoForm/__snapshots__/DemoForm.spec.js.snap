// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemoForm should call fieldChange when input value changed 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="demoForm"
>
    <h2>
        Sign up
    </h2>
    <div
        className="form-group "
    >
        <label
            htmlFor="email"
        >
            Email address
        </label>
        <input
            className="form-control"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="email"
            value=""
        />
        <p
            className="hidden"
        >
             is invalid
        </p>
    </div>
    <div
        className="form-group "
    >
        <label
            htmlFor="password"
        >
            Password
        </label>
        <input
            className="form-control"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
        />
        <p
            className="hidden"
        >
             is too short
        </p>
    </div>
    <button
        className="btn btn-primary"
        onClick={[Function]}
    >
        Sign up
    </button>
</div>,
  "nodes": Array [
    <div
      className="demoForm"
>
      <h2>
            Sign up
      </h2>
      <div
            className="form-group "
      >
            <label
                  htmlFor="email"
            >
                  Email address
            </label>
            <input
                  className="form-control"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="email"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is invalid
            </p>
      </div>
      <div
            className="form-group "
      >
            <label
                  htmlFor="password"
            >
                  Password
            </label>
            <input
                  className="form-control"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is too short
            </p>
      </div>
      <button
            className="btn btn-primary"
            onClick={[Function]}
      >
            Sign up
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DemoForm
        fieldChange={[Function]}
        formChange={[Function]}
        state={
                Object {
                        "email": "",
                        "emailValid": false,
                        "formErrors": Object {
                          "email": " is invalid",
                          "password": " is too short",
                        },
                        "formValid": false,
                        "password": "",
                        "passwordValid": false,
                        "touched": Object {
                          "email": false,
                          "password": false,
                        },
                      }
        }
/>,
      "_debugID": 13,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DemoForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSignUp": [Function],
        "props": Object {
          "fieldChange": [Function],
          "formChange": [Function],
          "state": Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
        "_debugID": 14,
        "_renderedOutput": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DemoForm
    fieldChange={[Function]}
    formChange={[Function]}
    state={
        Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          }
    }
/>,
}
`;

exports[`DemoForm should call formChange when buuton clicked 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="demoForm"
>
    <h2>
        Sign up
    </h2>
    <div
        className="form-group "
    >
        <label
            htmlFor="email"
        >
            Email address
        </label>
        <input
            className="form-control"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="email"
            value=""
        />
        <p
            className="hidden"
        >
             is invalid
        </p>
    </div>
    <div
        className="form-group "
    >
        <label
            htmlFor="password"
        >
            Password
        </label>
        <input
            className="form-control"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
        />
        <p
            className="hidden"
        >
             is too short
        </p>
    </div>
    <button
        className="btn btn-primary"
        onClick={[Function]}
    >
        Sign up
    </button>
</div>,
  "nodes": Array [
    <div
      className="demoForm"
>
      <h2>
            Sign up
      </h2>
      <div
            className="form-group "
      >
            <label
                  htmlFor="email"
            >
                  Email address
            </label>
            <input
                  className="form-control"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="email"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is invalid
            </p>
      </div>
      <div
            className="form-group "
      >
            <label
                  htmlFor="password"
            >
                  Password
            </label>
            <input
                  className="form-control"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is too short
            </p>
      </div>
      <button
            className="btn btn-primary"
            onClick={[Function]}
      >
            Sign up
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DemoForm
        fieldChange={[Function]}
        formChange={[Function]}
        state={
                Object {
                        "email": "",
                        "emailValid": false,
                        "formErrors": Object {
                          "email": " is invalid",
                          "password": " is too short",
                        },
                        "formValid": false,
                        "password": "",
                        "passwordValid": false,
                        "touched": Object {
                          "email": false,
                          "password": false,
                        },
                      }
        }
/>,
      "_debugID": 17,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DemoForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSignUp": [Function],
        "props": Object {
          "fieldChange": [Function],
          "formChange": [Function],
          "state": Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
        "_debugID": 18,
        "_renderedOutput": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DemoForm
    fieldChange={[Function]}
    formChange={[Function]}
    state={
        Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          }
    }
/>,
}
`;

exports[`DemoForm should call formChange when entering input 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="demoForm"
>
    <h2>
        Sign up
    </h2>
    <div
        className="form-group "
    >
        <label
            htmlFor="email"
        >
            Email address
        </label>
        <input
            className="form-control"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="email"
            value=""
        />
        <p
            className="hidden"
        >
             is invalid
        </p>
    </div>
    <div
        className="form-group "
    >
        <label
            htmlFor="password"
        >
            Password
        </label>
        <input
            className="form-control"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
        />
        <p
            className="hidden"
        >
             is too short
        </p>
    </div>
    <button
        className="btn btn-primary"
        onClick={[Function]}
    >
        Sign up
    </button>
</div>,
  "nodes": Array [
    <div
      className="demoForm"
>
      <h2>
            Sign up
      </h2>
      <div
            className="form-group "
      >
            <label
                  htmlFor="email"
            >
                  Email address
            </label>
            <input
                  className="form-control"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="email"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is invalid
            </p>
      </div>
      <div
            className="form-group "
      >
            <label
                  htmlFor="password"
            >
                  Password
            </label>
            <input
                  className="form-control"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value=""
            />
            <p
                  className="hidden"
            >
                   is too short
            </p>
      </div>
      <button
            className="btn btn-primary"
            onClick={[Function]}
      >
            Sign up
      </button>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DemoForm
        fieldChange={[Function]}
        formChange={[Function]}
        state={
                Object {
                        "email": "",
                        "emailValid": false,
                        "formErrors": Object {
                          "email": " is invalid",
                          "password": " is too short",
                        },
                        "formValid": false,
                        "password": "",
                        "passwordValid": false,
                        "touched": Object {
                          "email": false,
                          "password": false,
                        },
                      }
        }
/>,
      "_debugID": 15,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DemoForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSignUp": [Function],
        "props": Object {
          "fieldChange": [Function],
          "formChange": [Function],
          "state": Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
        "_debugID": 16,
        "_renderedOutput": <div
          className="demoForm"
>
          <h2>
                    Sign up
          </h2>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="email"
                    >
                              Email address
                    </label>
                    <input
                              className="form-control"
                              name="email"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="email"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is invalid
                    </p>
          </div>
          <div
                    className="form-group "
          >
                    <label
                              htmlFor="password"
                    >
                              Password
                    </label>
                    <input
                              className="form-control"
                              name="password"
                              onBlur={[Function]}
                              onChange={[Function]}
                              type="password"
                              value=""
                    />
                    <p
                              className="hidden"
                    >
                               is too short
                    </p>
          </div>
          <button
                    className="btn btn-primary"
                    onClick={[Function]}
          >
                    Sign up
          </button>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DemoForm
    fieldChange={[Function]}
    formChange={[Function]}
    state={
        Object {
            "email": "",
            "emailValid": false,
            "formErrors": Object {
              "email": " is invalid",
              "password": " is too short",
            },
            "formValid": false,
            "password": "",
            "passwordValid": false,
            "touched": Object {
              "email": false,
              "password": false,
            },
          }
    }
/>,
}
`;
